File: src/kiro_crew/dashboard/pinned_session_order.py
"""The person's order for pinned sidebar sessions, kept on the gateway.

One JSON file, ``chat_pinned_order.json`` in the config dir, holds a list of
slot keys. A key's position in the list is its rank. The list is global rather
than per folder: the sidebar renders pinned rows inside each folder (and at the
top level) and sorts them by rank there, so filtering the one list by group
gives each group's order. That is the same model the browser used when this
order lived in ``localStorage`` (``mc-pinned-session-order``), so the sidebar's
existing drag handling keeps working on the data it already understood.

The file is written whole with an atomic replace, so a reorder is one write:
it lands entirely or not at all.

Rules the writers follow:

* No file at all means the person never reordered. Pinned rows then keep the
  sidebar's own sort, which is what they had before this store existed. A
  file holding an empty list is an order that was saved and has since emptied.
* Pinning a session appends it to the end once an order has been saved, even
  one that has since emptied. A person who never reordered keeps the plain
  sort.
* Unpinning a session removes it, so re-pinning later puts it at the end
  instead of back in an old slot.
* A key whose slot is gone or is not pinned has no rank. Such keys are
  dropped on the next write.
"""

from __future__ import annotations

import json
import logging
from collections.abc import Iterable, Mapping
from pathlib import Path
from typing import Any

from kiro_crew.atomic_write import atomic_write

logger = logging.getLogger(__name__)

PINNED_ORDER_FILE = "chat_pinned_order.json"

#: The most keys the store keeps and one reorder request may carry. Far above
#: any real sidebar; a request past it is malformed rather than large.
MAX_PINNED_ORDER_KEYS = 2000

#: Slot keys are short generated identifiers; this bounds a hand-edited file
#: and a request body without constraining any real key.
MAX_PINNED_ORDER_KEY_CHARS = 256


def _valid_key(value: Any) -> bool:
    return isinstance(value, str) and 0 < len(value) <= MAX_PINNED_ORDER_KEY_CHARS


def load(path: Path) -> list[str]:
    """Read the stored order from *path*.

    A missing file is the normal first-run state and reads as ``[]``. A file
    that does not parse, or is not a list, also reads as ``[]`` with a warning:
    the order is a display preference, so losing it degrades to the sidebar's
    own sort rather than failing startup. Entries that are not usable keys and
    repeated keys are skipped. A transient ``OSError`` is raised, so a caller
    never replaces good in-memory state with an empty read.
    """
    try:
        if not path.exists():
            return []
        raw = json.loads(path.read_text(encoding="utf-8"))
    except (UnicodeDecodeError, ValueError) as exc:
        logger.warning("%s has malformed content: %s", PINNED_ORDER_FILE, exc)
        return []
    if not isinstance(raw, list):
        logger.warning("%s is not a list (%s); ignoring", PINNED_ORDER_FILE, type(raw).__name__)
        return []
    return dedupe(key for key in raw if _valid_key(key))[:MAX_PINNED_ORDER_KEYS]


def save(path: Path, keys: list[str]) -> None:
    """Replace the stored order with *keys* in one atomic, owner-only write."""
    atomic_write(path, json.dumps(keys), fsync=True, mode=0o600)


def dedupe(keys: Iterable[str]) -> list[str]:
    """*keys* in order with later repeats removed."""
    seen: set[str] = set()
    out: list[str] = []
    for key in keys:
        if key not in seen:
            seen.add(key)
            out.append(key)
    return out


def prune(keys: Iterable[str], pinned: Mapping[str, bool]) -> list[str]:
    """Keep only keys whose slot is live and pinned, in order.

    *pinned* maps each live slot key to its pinned flag.
    """
    return [key for key in dedupe(keys) if pinned.get(key)]


def after_pin_change(
    current: list[str],
    key: str,
    now_pinned: bool,
    pinned: Mapping[str, bool],
    stored: bool,
) -> list[str]:
    """The order after *key* is pinned or unpinned.

    Unpinning removes the key. Pinning appends it once the person has ordered
    their pins at all (*stored*: an order was saved, even one that has since
    emptied). A person who never reordered keeps the plain sort.
    """
    rest = [k for k in current if k != key]
    if now_pinned and (stored or rest):
        rest.append(key)
    return prune(rest, pinned)


def after_reorder(
    current: list[str], requested: list[str], pinned: Mapping[str, bool]
) -> list[str]:
    """The order after a reorder that puts *requested* first, in that order.

    Stored keys the request did not name keep their relative order after it,
    so a caller that ranks only part of the pinned set does not erase the rest.
    """
    named = set(requested)
    return prune([*requested, *(k for k in current if k not in named)], pinned)

File: website/src/utils/pinnedSessionOrder.ts
import { safeGetItem, safeRemoveItem } from './safeStorage'

/**
 * The order of pinned sessions lives on the gateway (`chat_pinned_order.json`,
 * written by `POST /api/chat/pinned-order`). Each slot row carries its place as
 * `pin_rank`: an index into one global list, or `null` when the session has no
 * stored place. Every surface ranks pinned rows from that field, so the sidebar,
 * the flyout, the Sessions page and every browser the person opens agree.
 *
 * Pinned rows without a rank follow the ranked ones in the surface's own sort.
 * Before any reorder no row has a rank, so the pinned section keeps its plain
 * sort.
 */

/**
 * Where this browser kept the order before the gateway did. Read once, handed
 * to the gateway, then removed (see ChatSidebar's migration effect).
 */
export const LEGACY_PINNED_SESSION_ORDER_KEY = 'mc-pinned-session-order'

/** The legacy browser-local order. Invalid or unavailable storage reads as none. */
export function readLegacyPinnedSessionOrder(): string[] {
  try {
    const parsed: unknown = JSON.parse(safeGetItem(LEGACY_PINNED_SESSION_ORDER_KEY) || '[]')
    if (!Array.isArray(parsed)) return []
    return parsed.filter((key): key is string => typeof key === 'string')
  } catch {
    return []
  }
}

export function clearLegacyPinnedSessionOrder(): void {
  safeRemoveItem(LEGACY_PINNED_SESSION_ORDER_KEY)
}

interface RankedRow {
  key: string
  pinned?: boolean
  pin_rank?: number | null
}

/** Keys of pinned rows that have a stored place, in that order. */
export function rankedPinnedKeys(rows: readonly RankedRow[]): string[] {
  return rows
    .filter((row): row is RankedRow & { pin_rank: number } => !!row.pinned && typeof row.pin_rank === 'number')
    .sort((a, b) => a.pin_rank - b.pin_rank)
    .map(row => row.key)
}

/**
 * Keep stored keys that are still pinned, discard duplicates/stale keys, then
 * append the remaining pinned sessions in the caller's natural sort order.
 */
export function reconcilePinnedSessionOrder(
  stored: readonly string[],
  natural: readonly string[],
): string[] {
  const valid = new Set(natural)
  const seen = new Set<string>()
  const out: string[] = []
  for (const key of stored) {
    if (valid.has(key) && !seen.has(key)) {
      seen.add(key)
      out.push(key)
    }
  }
  for (const key of natural) {
    if (!seen.has(key)) {
      seen.add(key)
      out.push(key)
    }
  }
  return out
}

/** Move one pinned key to another key's position. Unknown/equal keys are inert. */
export function movePinnedSession(
  order: readonly string[],
  activeKey: string,
  overKey: string,
): string[] {
  const from = order.indexOf(activeKey)
  const to = order.indexOf(overKey)
  if (from < 0 || to < 0 || from === to) return [...order]
  const next = [...order]
  const [moved] = next.splice(from, 1)
  next.splice(to, 0, moved)
  return next
}
